File: IT/1/Oggetti/Esercizi/Python/esercizio22.py
# file esercizio22.py

from __future__ import print_function

# la classe Cat rappresenta un gatto
class Cat(object):
	def __init__(self):
		self.__spec = "gatto"

	def __str__(self):
		return self.__spec

# la classe Dog rappresenta un cane 
class Dog(object):
	def __init__(self):
		self.__spec = "cane"

	def __str__(self):
		return self.__spec

# la classe Cow rappresenta una mucca
class Cow(object):
	def __init__(self):
		self.__spec = "mucca"

	def __str__(self):
		return self.__spec

class Factory(object):
	def getAnimal(self, animal):
		if (animal == "gatto"):
			return Cat()

		if (animal == "cane"):
			return Dog()

		if (animal == "mucca"):
			return Cow()

		return None

if __name__ == "__main__":
	factory = Factory()

	animal = factory.getAnimal("gatto")
	print(animal)

	animal = factory.getAnimal("cane")
	print(animal)

	animal = factory.getAnimal("mucca")
	print(animal)

File: IT/1/Oggetti/Esercizi/Java/Esercizio22.java
/* file Esercizio22.java */

/* l'interfaccia Animal identifica un animale generico */

interface Animal {
	public abstract void spec();
}

/* la classe Cat rappresenta un gatto */
class Cat implements Animal {
	@Override
	public void spec() {
		System.out.println("gatto");
	}
}

/* la classe Dog rappresenta un cane */
class Dog implements Animal {
	@Override
	public void spec() {
		System.out.println("cane");
	}
}

/* la classe Cow rappresenta una mucca */
class Cow implements Animal {
	@Override
	public void spec() {
		System.out.println("mucca");
	}
}

class Factory {
	public Animal getAnimal(String animal) {
		if ("gatto".equals(animal)) {
			return new Cat();
		}
		if ("cane".equals(animal)) {
			return new Dog();
		}
		if ("mucca".equals(animal)) {
			return new Cow();
		}

		return null;
	}
}

public class Esercizio22 {
	public static void main(String [] args) {
		Factory factory = new Factory();
		Animal animal = null;

		animal = factory.getAnimal("gatto");
		animal.spec();

		animal = factory.getAnimal("cane");
		animal.spec();

		animal = factory.getAnimal("mucca");
		animal.spec();
	}
}

